src/azure-cli-core/azure/cli/core/keys.py

```diff
@@ -37,9 +41,17 @@ def generate_ssh_keys(private_key_filepath, public_key_filepath):
         os.makedirs(ssh_dir)
         os.chmod(ssh_dir, 0o700)
 
-    key = paramiko.RSAKey.generate(2048)
-    key.write_private_key_file(private_key_filepath)
-    os.chmod(private_key_filepath, 0o600)
+    try:
+        # try to use existing private key if it exists.
+        key = paramiko.RSAKey(filename=private_key_filepath)
+        logger.warning("Private SSH key file '%s' found in dir '%s'."
+                       " Generating public key file '%s'",
+                       private_key_filepath, ssh_dir, public_key_filepath)
+    except:
+        # otherwise generate new private key.
+        key = paramiko.RSAKey.generate(2048)
+        key.write_private_key_file(private_key_filepath)
+        os.chmod(private_key_filepath, 0o600)
 
     with open(public_key_filepath, 'w') as public_key_file:
```

Comment:
I believe the public key file will get overwritten as well. Should that also be fixed?

Reply:
I left it as is. My rationale was that in the event a public key path / value is given, `validate_ssh_key` (in the acs and vm modules) is called only if the public key filepath or value given does not exist or is invalid. If there was already a correct public key file, `validate_ssh_key` will not be called in the first place.

The code that makes this check starts at line 867 of vm's `_validators.py`.

```
def validate_ssh_key(namespace):
    string_or_file = (namespace.ssh_key_value or
                      os.path.join(os.path.expanduser('~'), '.ssh', 'id_rsa.pub'))
    content = string_or_file
>>> if os.path.exists(string_or_file):
        logger.info('Reusing existing SSH public key from %s', string_or_file)
        with open(string_or_file, 'r') as f:
            content = f.read() 
>>> elif not keys.is_valid_ssh_rsa_public_key(content): 
        if namespace.generate_ssh_keys:
            # figure out appropriate file names:
            # 'base_name'(with private keys), and 'base_name.pub'(with public keys)
            public_key_filepath = string_or_file
            if public_key_filepath[-4:].lower() == '.pub':
                private_key_filepath = public_key_filepath[:-4]
            else:
                private_key_filepath = public_key_filepath + '.private'
            content = keys.generate_ssh_keys(private_key_filepath, public_key_filepath)
```

Reply:
`generate_ssh_keys` is part of core and can be used by any module. The functionality in the method should not depend on checks in every module it is used in.

Reply:
Okay, this is fair enough. I will try to handle the 4 different cases of whether each of the keys exist.

Reply:
I made some changes. Let me know your thoughts.